forecasttools/utils.py

```diff
@@ -0,0 +1,183 @@
+                f"Parameter '{param_name}' must be one of the types {expected_type}; got {type(value)}"
+            )
+    else:
+        if not isinstance(value, expected_type):
+            raise TypeError(
+                f"Parameter '{param_name}' must be of type '{expected_type.__name__}'; got {type(value)}"
+            )
+
+
+def validate_and_get_start_time(start_date_iso: any):
+    """Handles the start_date_iso input,
+    converting it to a datetime object.
+
+    If the input is a string, it must be in
+    the format 'YYYY-MM-DD' or
+    'YYYY-MM-DD HH:MM:SS'. If it's a
+    datetime object, it is returned as is.
+    Otherwise, a TypeError is raised.
+    """
+    if isinstance(start_date_iso, str):
+        # try parsing both formats
+        for fmt in ("%Y-%m-%d", "%Y-%m-%d %H:%M:%S"):
+            try:
+                return datetime.strptime(start_date_iso, fmt)
+            except ValueError:
+                continue
+        # raise error if neither format matches
+        raise ValueError(
+            f"Parameter 'start_date_iso' must be in the format 'YYYY-MM-DD' or 'YYYY-MM-DD HH:MM:SS' if provided as a string; got {start_date_iso}."
+        )
+    elif isinstance(start_date_iso, datetime):
+        return start_date_iso
+    else:
+        raise TypeError(
+            f"Parameter 'start_date_iso' must be of type 'str' or 'datetime'; got {type(start_date_iso)}."
+        )
+
+
+def validate_and_get_idata_group(idata: az.InferenceData, group: str):
+    """Retrieves the group from the
+    InferenceData object and validates its
+    existence."""
+    idata_group = getattr(idata, group, None)
+    if idata_group is None:
+        raise ValueError(f"Group '{group}' not found in idata object.")
+    return idata_group
+
+
+def validate_and_get_idata_group_var(
+    idata_group: xr.Dataset,
+    group: str,
+    variable: str,
+):
+    """Retrieves the variable from the group
+    and validates its existence."""
+    if variable not in idata_group.data_vars:
+        raise ValueError(
+            f"Variable '{variable}' not found in group '{group}'."
+        )
+    return idata_group[variable]
+
+
+def validate_idata_group_var_dim(variable_data: xr.DataArray, dimension: str):
+    """Validates if the given dimension
+    exists in the variable."""
+    if dimension not in variable_data.dims:
+        raise ValueError(
+            f"Dimension '{dimension}' not found in variable dimensions: '{variable_data.dims}'."
+        )
+
+
+def validate_group_var_dim_instances(
+    iterable: Iterable, expected_type: type, param_name: str
+):
+    """
+    Check if all elements in the iterable are
+    instances of expected_type. If not, raise
+     a TypeError.
+    """
+    if not all(isinstance(item, expected_type) for item in iterable):
+        raise TypeError(
+            f"All items in '{param_name}' must be of type '{expected_type.__name__}'."
+        )
+
+
+def generate_time_range_for_dim(
```

Comment:
If it's specific to an idata dimension, which it currently is, this belongs in `idata_w_tdates_to_df`, not `utils`